### PR TITLE
refactor: compiling kcl code to object file instead of dylib and packaging LLVM lld library statically into kclvm binary.

### PR DESCRIPTION
**Pull request by Peefy**
#### 1. Does this PR affect any open issues?(Y/N) and add issue references (e.g. "fix #123", "re #123".):

- [ ] N
- [x] Y 

re #53. 

In order to better package KCLVM and reduce the probability of errors in the user's local kcl binary, the llvm library is statically packaged into the kcl binary to avoid using the user's local clang for assembling and linking. The goal is to solve the instability problem when KCL uses the user's clang/ld for code linking (mainly due to different versions of built-in cc on different OS).

Besides, after solving this problem, it is necessary to further move the python3 included in the KCLM from the KCLVM installation package to the outside to reduce the final KCLVM installation package size. cc @chai2010 

##### Design

The previous compilation process is:
image

After the issue is completed, the KCLVM binary package only contains the following contents (Take the macos target as an example, and the size is ~80M-90M):

* kclvm_cli  # KCLVM Rust CLI (using `kc run` or `kpm run` later.)
* libkclvm_native.dylib  # KCLVM Runtime native Lib for native targets including macos, linux and windows target.
* libkclvm_wasm32.dylib  # KCLVM WASM32 lib for the WASM target.

Note: 

* Support targets: mac-x86, mac-arm, linux-x86 target.
* TODO targets: windows, linux-arm  WASM target.

#### 2. What is the scope of this PR (e.g. component or file name):

* `kclvm_compiler` crate.
* `kclvm_runner` crate.

#### 3. Provide a description of the PR(e.g. more details, effects, motivations or doc link):

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [x] Other

#### 4. Are there any breaking changes?(Y/N) and describe the breaking changes(e.g. more details, motivations or doc link):

- [x] N
- [ ] Y 

#### 5. Are there test cases for these changes?(Y/N) select and add more details, references or doc links:

- [x] Unit test
- [x] Integration test
- [ ] Benchmark (add benchmark stats below)
- [ ] Manual test (add detailed scripts or steps below)
- [ ] Other

* Including all grammar and unit test suites.
* Add macos github runners on macos10, macos11, macos12.

#### 6. Release note

Please refer to [Release Notes Language Style Guide](https://kusionstack.io/docs/governance/release-policy/) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: kclvm/runner/src/assembler.rs [hunk 1]
```diff
@@ -320,11 +302,15 @@ impl KclvmAssembler {
             let tx = tx.clone();
             // clone a single file assembler for one thread.
             let assembler = self.single_file_assembler.clone();
-
-            let code_file = self.entry_file.clone();
+            let is_main_pkg = pkgpath == kclvm_ast::MAIN_PKG;
```
Comment (ldxdl) [inline kclvm/runner/src/assembler.rs:305]: pls leave a comment on what changed

File: .github/workflows/ubuntu_test.yaml [hunk 2]
```diff
@@ -2,10 +2,11 @@ name: build-and-test-ubuntu
 on: ["push", "pull_request"]
 jobs:
   build-and-test:
+    # Ref: https://github.com/actions/runner-images/tree/main/images/linux
     name: Test
     runs-on: ubuntu-latest
     container: 
-      image: kusionstack/kclvm-builder
+      image: peefyxpf/kclvm-builder
```
Comment (ldxdl) [inline .github/workflows/ubuntu_test.yaml:9]: Revert it

File: kclvm/runner/build.rs [hunk 3]
```diff
@@ -0,0 +1,66 @@
+fn main() {
```
Comment (ldxdl) [inline kclvm/runner/build.rs:1]: Pls leave a source info about files you copied.

File: kclvm/runner/src/assembler.rs [hunk 4]
```diff
@@ -37,8 +36,8 @@ pub(crate) trait LibAssembler {
     /// e.g. LLVM IR -> return : ".ll"
```
Comment (ldxdl) [inline kclvm/runner/src/assembler.rs:36]: Change .ll to .o.
Do a global search on it.

File: kclvm/runner/src/assembler.rs (hunk 1 above)
Comment (Peefy) [reply kclvm/runner/src/assembler.rs:305]: > pls leave a comment on what changed

Same with #272, I have reverted it.

File: .github/workflows/ubuntu_test.yaml (hunk 2 above)
Comment (Peefy) [reply .github/workflows/ubuntu_test.yaml]: > Revert it

done

File: kclvm/runner/build.rs (hunk 3 above)
Comment (Peefy) [reply kclvm/runner/build.rs:6]: > Pls leave a source info about files you copied.

Have added documents to explain it.

File: kclvm/runner/src/assembler.rs (hunk 4 above)
Comment (Peefy) [reply kclvm/runner/src/assembler.rs]: > Change .ll to .o. Do a global search on it.

fixed.

**Comment by ldxdl** [comment]
lgtm

**Review by ldxdl — DISMISSED** [review]

**Review by zong-zhe — APPROVED** [review]
LGTM

**Review by ldxdl — APPROVED** [review]

**Pull request merged**